### ISSUE TITLE
Icy title is not displayed until next song starts playing after changing volume.

### ISSUE THREAD

**Issue by crocket** [coderholic/pyradio#62]
After changing volume, the status bar at the bottom displays the volume until the next song starts playing.

I really want to see icy title... Icy title should be always displayed.

I want the status bar to look like

```
icy-title: BlahBlah - SDFSDF
Volume 33%
```

**Comment by s-n-g** [comment]
Working on it....

**Comment by s-n-g** [comment]
Ok, done

Considered Closed

**Comment by crocket** [comment]
Thanks.

**Comment by s-n-g** [comment]
Reopening, since we reverted to the time before this issue was opened

**Comment by crocket** [comment]
I executed the following commands.

```
git clone https://github.com/s-n-g/pyradio.git
git checkout threaded-delay-local
python
from pyradio.main import shell
shell()
```

For Anime Nfo, icy title is not displayed until I change volume.

**Comment by s-n-g** [comment]
Yes, I know, I have already fixed that :)
I would suppose the volume delay threading thingie is ok now for you too? you said nothing about it...

Thanks for sending me the link; is it the one you get titles all screwed up?

BTW, what backend do you use? (mpv, mplayer, vlc?)

And FYI, I find myself listening to this (most of the time)

```
The Original 1940s Radio Station,http://yp.shoutcast.com/sbin/tunein-station.pls?id=959067
```
:)

**Comment by crocket** [comment]
I listen to http://www.mukulcast.com, the icy title charset of which is CP949.

**Comment by crocket** [comment]
This issue is fixed. But, icy title is not displayed until I change the volume when I listen to Anime Nfo Radio or mukulcast.

**Comment by s-n-g** [comment]
Maybe you should update with git again
I have also changed mpv and mplayer title and volume display strings
For me, everything seems ok, both on manjaro and mx linux

BTW, that radio station, Anim Nfo, what great music it plays!!!
It's in my favorites now. Thanks for that!

**Comment by s-n-g** [comment]
Look at his screenshot. Upper part is the log file, lower part is the player
Is that how it's supposed to be written?

Also look at the icy string
![_2018-11-13_15-57-22](https://user-images.githubusercontent.com/5807638/48427318-27cf0700-e771-11e8-90da-dcab16e3de0c.png)

**Comment by crocket** [comment]
That looks like a valid kanji character used in japan.

**Comment by s-n-g** [comment]
yes, but it is displayed instead of an "H"...
anyway, it seems utf is ok here... or not?

**Comment by crocket** [comment]
I am not sure.

**Comment by crocket** [comment]
![2018-11-13_tue_16 22 40](https://user-images.githubusercontent.com/748856/48427897-27ecf800-e7ad-11e8-811a-293176cd6f25.png)

As you can see in the above screenshot, icy title is not displayed until I change volume.

I tested it on c37cdae12e91056f417d042a3a38fce49cdca9cf

**Comment by s-n-g** [comment]
![anim](https://user-images.githubusercontent.com/5807638/48428354-81383580-e773-11e8-8c21-57774eb9a672.gif)

Running pyradio -d. I typed nothing after the player starts playing.
When the server sends the icy-title, I get it on the screen

**Comment by crocket** [comment]
It seems icy-title is not fetched early for AnimeNfo. But, changing volume hastens the fetching process.

Mukulcast seems to never fetch icy title until I change volume.

**Comment by s-n-g** [comment]
I guess you just have to wait...

Mukulcat, I can't play that...
![anim](https://user-images.githubusercontent.com/5807638/48429117-2a336000-e775-11e8-84a8-48278359cbb4.gif)

**Comment by crocket** [comment]
http://www.mukulcast.com/listen.pls is encoded in CP949. You have to convert the character encoding of listen.pls to UTF-8.

**Comment by crocket** [comment]
I tested https://github.com/s-n-g/pyradio/commit/7bd91dff2d9d9ea6be215a03a1b831816d4df9ae

`pyradio -d` reveals that icy title is displayed when it is fetched for the second time.

For the first time, it is not displayed.

If I change the volume, the firstly fetched icy title is displayed.

**Comment by s-n-g** [comment]
```
LANG=en_US pacman -Qi python
Name            : python
Version         : 3.7.1-1
Description     : Next generation of the python high-level scripting language
Architecture    : x86_64
URL             : http://www.python.org/
Licenses        : custom
Groups          : None
Provides        : python3
Depends On      : expat  bzip2  gdbm  openssl  libffi  zlib  libnsl
Optional Deps   : python-setuptools [installed]
                  python-pip [installed]
                  sqlite [installed]
                  mpdecimal: for decimal
                  xz: for lzma [installed]
                  tk: for tkinter [installed]
Required By     : autojump  bind-tools  fontforge  gdb-common  geany  geany-plugins  gobject-introspection  gufw  iotop  libixion  libopenshot  libreoffice-fresh  menulibre  menumaker  mono
                  mps-youtube  namcap  nuitka  ob-autostart  pacman-mirrors  pclcmd  pcsclite  pyalpm  pyradio-git  python-appdirs  python-asn1crypto  python-attrs  python-beaker
                  python-beautifulsoup4  python-cairo  python-click  python-colorama  python-dbus  python-diff-match-patch  python-distlib  python-distro  python-distutils-extra
                  python-entrypoints  python-gobject  python-greenlet  python-html5lib  python-httplib2  python-idna  python-jedi  python-jeepney  python-libtmux  python-lockfile
                  python-lxml  python-markupsafe  python-minidb  python-msgpack  python-nose  python-npyscreen  python-olefile  python-pafy  python-parso  python-pexpect  python-pillow
                  python-ply  python-polib  python-powerline  python-progress  python-progressbar  python-psutil  python-ptyprocess  python-pycryptodome  python-pycups  python-pycurl
                  python-pyelftools  python-pyparsing  python-pypdf2  python-pypeg2  python-pytoml  python-pyzmq  python-robobrowser  python-rope  python-selenium  python-setproctitle
                  python-sip  python-sip-pyqt5  python-six  python-urllib3  python-utils  python-webencodings  python-werkzeug  python-xapp  python-xdg  python-yaml  ranger-git  setconf  ufw
                  vim-jedi  virtualbox  youtube-dl
Optional For    : boost  btrfs-progs  fish  glib2  gvim  hexchat  ki18n  lensfun  libftdi  libnewt  libproxy  lirc  nfs-utils  refind-efi  texlive-core  usbutils  util-linux  vim-runtime
                  volume_key
Conflicts With  : None
Replaces        : python3
Installed Size  : 145.31 MiB
Packager        : Felix Yan <felixonmars@archlinux.org>
Build Date      : Mon Oct 22 13:41:28 2018
Install Date    : Fri Oct 26 13:19:43 2018
Install Reason  : Installed as a dependency for another package
Install Script  : No
Validated By    : Signature
```

I am on manjaro linux

Can you please send your pyradio.log?
Just clear it, make a connection to Mukulcast (pyradio -d) and exit
Unfortunately, I cannot connect to Mukulcast

**Comment by crocket** [comment]
[pyradio.log](https://github.com/coderholic/pyradio/files/2580435/pyradio.log)

**Comment by s-n-g** [comment]
ok, mukilcast does not work for me, no matter what...
but... maybe I have a solution

Please edit pyradio/player.py, line 155, and change
```
if self.icy_found and (self.PLAYER_CMD == 'mplayer'):
```
to
```
if self.icy_found:
```

Then rebuild and test

**Comment by crocket** [comment]
Changing

```
if self.icy_found and (self.PLAYER_CMD == 'mplayer'):
```

to

```
if self.icy_found:
```

works.

I think pyradio is far more complicated than it can be. If it could become simple, it should be. Why don't you set the format of the status bar as follows?

```
Vol: 35%[, Icy-Title: blah blah ...]
```

The status bar doesn't really have to show debug information. This is going to simplify implementation a lot.

**Comment by s-n-g** [comment]
Well, that may be a way to go...
I will have to look into it, but the idea here is that one should be able to see server messages until the connection is done (i.e. get the first icy-title, which means that the player is good to go)

It will take some thinking... but thank you very much for the suggestion - and all the help you have provided till now

**Comment by crocket** [comment]
You can show debug information on the second line of the status bar as below.

```
Vol: 35%[ Icy-title: blah blah ...]
ALSA blah blah ....
```

With that layout, implementation is far simpler. Do you have to keep the user interface compact because you run pyradio in small windows?

**Comment by s-n-g** [comment]
Well, that is a drastic change, and it's not me who can decide about it, I am just helping out...
After all, this is @coderholic 's project :)

**Comment by coderholic** [comment]
Yeah I don't think we want to make the status bar bigger just because you sometimes need to wait for a title :) 

We could right align the volume when it changes perhaps, eg.

```
TITLE                           Vol: 35%
```

But it's been a while since I looked at the code so I'm can't even remember how easy or not that'd be. Another thing we could do it set the text back to whatever it was after a second or something.

So...

```
current_text = get_status()
volume_status = "Vol: 35%"
set_status(volume_status)
sleep(2)
if get_status() == volume_status: # make sure it didn't change, eg. new song
  set_status(current_text)
```

**Comment by s-n-g** [comment]
I just made a PR
At this point the status is as it was before, but

1. when volume changes/get saved, it is displayed for 1 sec and then we get back to whatever text we had before

2. after we get the first icy-title, we only display icy-title (whenever the next one comes)

**Comment by coderholic** [comment]
Awesome, that sounds like the perfect solution! :)

**Comment by s-n-g** [comment]
You will see I really screwed up while trying to get to it :)
I had to revert to the previous "last good" condition and start over again...

I was thinking of what @crocket said, to have something like 

```
[ Vol: xx%] icy-title
```

but the thing is, we do not know the volume value before we change it (we get it from program response), so it's a bit tricky...

Ok, I will leave you to it, you will want to check this one :)

**Comment by crocket** [comment]
```
Yeah I don't think we want to make the status bar bigger just because you sometimes need to wait for a title :)
```

The point is not to eliminate waiting, but to simplify implementation in order to prevent bugs from occurring. Using threads to coordinate information display on a single line is going to be fragile.

You can also assign a keyboard shortcut for toggling the debug information bar.

**Comment by coderholic** [comment]
I'm fine with the goal of simplifying the implementation, but simplifying the UI is a more important goal - I'd rather messier internals than messier UI!

**Comment by crocket** [comment]
What about using [sched queue](https://docs.python.org/3/library/sched.html) or a queue? Scheduling events in a queue feels safer than another thread.

**Comment by s-n-g** [comment]
About sched:

From [2.7.15 documentation](https://docs.python.org/2/library/sched.html):
> In multi-threaded environments, the scheduler class has limitations with respect to thread-safety, inability to insert a new task before the one currently pending in a running scheduler, and holding up the main thread until the event queue is empty. Instead, the preferred approach is to use the threading.Timer class instead.

From [3.7.1 documentation](https://docs.python.org/3/library/sched.html):
> Changed in version 3.3: scheduler class can be safely used in multi-threaded environments.

**Comment by crocket** [comment]
I figured out that the right abstraction to use depends on what you want to do. Since I don't know what exactly you want to do, it's hard to know what the right abstraction is.

**Comment by s-n-g** [comment]
:)
Well, this is what I want to do:
pyradio behavior:
1. user starts to play a station
2. pyradio connects to the station - I want all messages to be shown to the user, until the connection is established - I know that this has happened when I get the first **icy-title**
3. Once I get the **first icy-title**, display **only icy-title** messages (i.e. when song changes)

When volume changed/saved:
1. display current value for 1 sec (or messages "volume saved", etc.)
2. after 1 sec, display whatever was previously on the status bar (either a server message or a song title)
3. when repeatedly changing the volume (i.e. two or more changes within 1 sec), only show volume value (i.e. not having the title popping up)

**Comment by s-n-g** [comment]
Yes, a single thread is just fine.
And yes, I did struggle with threading, and I did pull it off
And yes, again, messiness is subjective. I, on the other hand, like a clean UI.

If @coderholic likes this approach, he will go with it
Otherwise, we'll work something else out

**Comment by coderholic** [comment]
I don't write much python these days, but this SO post seems to have some great solutions for the delay: https://stackoverflow.com/questions/10154568/postpone-code-for-later-execution-in-python-like-settimeout-in-javascript

**Comment by crocket** [comment]
I tested the latest commit from s-n-g/pyradio. It seems to fix this issue.

**Comment by s-n-g** [comment]
That's what I've been trying to tell you the last couple of days...
This issue has been fixed since my PR went on
Now I'm just working on another feature on s-n-g/pyradio...

@coderholic the link you've provided, says:
1. use **threading.Timer** (which we do) and if that's not enough use priority queue. In our case, we already know the priority... **display volume for 1 sec, no matter what** (as if we have a tremendous amount of data to display :) )
2. use signals on linux - for cross platform apps, use **threading.Timer** :)
3. use **heapq** - but this is not the behaviour we want here; we want to display the last volume message, wait 1 sec and restore whatever was on the status bar before we started displaying volume messages **and** cancel any other volume messages (threads) waiting to restore their message
4. finally, use **asyncio**. why use it here? we are already multi-threading, why go asyncio too, for such a simple thing?

I think this code is "production ready" and full-proof at this point, but I will wait for you to accept or reject it :)

I also have the "-u, --use-player command line ready", the ability to choose which player we want to use, but let's get this one on or off first
